rest/oidc_api.go

```diff
@@ -99,9 +108,38 @@ func (h *handler) handleOIDCCommon() (redirectURLString string, err error) {
 		return redirectURLString, err
 	}
 
+	if !provider.IsDefault {
+		base.Debugf(base.KeyAuth, "Adding provider (%v) to callback URL", base.UD(provider.Name))
+		if err = addCallbackURLQueryParam(redirectURL, requestParamProvider, provider.Name); err != nil {
+			base.Errorf("Failed to add provider to callback URL, err: %v", err)
+		}
+		base.Debugf(base.KeyAuth, "Callback URL: %s", redirectURL.String())
+	}
+
 	return redirectURL.String(), nil
 }
 
+func addCallbackURLQueryParam(uri *url.URL, name, value string) error {
+	if uri == nil {
+		return errors.New("URL must not be nil")
+	}
+	if name == "" {
+		return errors.New("parameter name must not be empty")
+	}
+	rawQuery, err := url.ParseQuery(uri.RawQuery)
+	if err != nil {
+		return err
+	}
+	redirectURL := rawQuery.Get(requestParamRedirectURI)
+	if redirectURL == "" {
+		return errors.New("no " + requestParamRedirectURI + " parameter found in URL")
+	}
+	redirectURL += "&" + name + "=" + value
+	rawQuery.Set(requestParamRedirectURI, redirectURL)
```

Comment:
What about when the URL of `redirect_uri` itself has no query parameters set? We'd need to append `?name=value` instead of `&name=value`. It looks like all of the test cases follow this pattern and are actually checking for the wrong result.

Would it be best to parse `redirect_uri` using url.Parse and set a query parameter using the strongly typed Query/Value methods?